### ISSUE TITLE
About Deformable Attention

### ISSUE THREAD

**Issue by Tsai-chia-hsiang** [helq2612/BiADT#3]
Hi, I would like to ask about the Deformable Attention mechanism in the paper.

I went to the paper DEFORMABLE DETR: DEFORMABLE TRANSFORMERS
FOR END-TO-END OBJECT DETECTION and the Deformable Attention 
compute different sample k for each query.

However, I see that it directly use $A_I\times V_I$ and it seems like all query using the same sampling $n'$ result. 

I am a little bit confused about that. 

Thank you for your answer.

**Comment by helq2612** [comment]
I don't think all the queries are using the same sampling results. The points (offset) sampling is the same as the one in the original deformable attention: using a linear projection to generate offsets for each query, see [line 124](https://github.com/helq2612/BiADT/blob/d6ab363681b00706801f3f720d37da33de4f04d9/models/dn_dab_deformable_detr/ops/modules/ms_deform_attn.py#L124C50-L124C62).

**Comment by Tsai-chia-hsiang** [comment]
Thanks for the reply.  I've seen the code. 
But what I really want to ask is the mathematical equation in the p.4  eq(5) of the paper.

the formula: $\textbf{DeformAttn}(C_I, P_I, Y_I)=[A_I\cdot V_I]W^{o}_I$ (5)

if $A_I\in R^{n\times n'}$ and $V_I\in R^{n'\times 2d}$., where the $\cdot$ operator I assume that it is the conventional matrix multiplication : 
(take $n'=4$ as an example)

$$A_I = \begin{bmatrix}
    \textbf{query} 1 = a_{q_{10}} & a_{q_{11}} & a_{q_{12}} & a_{q_{13}} \\
   ... & ... & ... & ... & \\
    \textbf{query} n = a_{q_{n0}} & a_{q_{n1}} & a_{q_{n2}} & a_{q_{n3}}
\end{bmatrix}_{n\times 4}$$

$$V_I=\begin{bmatrix}
   v_{1d_1} & v_{1d_2}  & ... &  v_{1d_{256}} \\
  v_{2d_1} & v_{2d_2}  & ... &  v_{2d_{256}} \\
  v_{3d_1} & v_{3d_2}  & ... &  v_{3d_{256}} \\
  v_{4d_1} & v_{4d_2}  & ... &  v_{4d_{256}}
\end{bmatrix}_{4 \times 2d}$$

and  $A_I \times V_I=$

$$\begin{bmatrix}
a_{\textbf{query1}} \cdot V_I \\
... \\
a_{\textbf{queryn}} \cdot V_I \\
\end{bmatrix}_{n\times 2d}$$

and seems like each query $\in{1,2,... ,n}$ just get attention from the same $V_I$ ?

Thanks.

**Comment by helq2612** [comment]
Yes, the same context V.

**Comment by Tsai-chia-hsiang** [comment]
But shouldn't it be different $V_I$ for different query $q\in 1,2,...,n$ ?

The origin equation of one head $m$ deformable attention from DEFORMABLE DETR: DEFORMABLE TRANSFORMERS
FOR END-TO-END OBJECT DETECTION : 

$\displaystyle\sum_{k=1}^{K}A_{mqk}W_m'x(p_q+\Delta_{p_{mqk}})$, The $V_I$ should be different.

$$\textbf{query} q  = \begin{bmatrix}
    a_{q_{0}} & a_{q_{1}} & a_{q_{2}} & a_{q_{3}}
\end{bmatrix}_{1\times 4}$$ 

get attention for a unique $V_{I_q}\in R^{n'\times 2d}$

i.e. $V_I$ should has the shape like $(n \times n'\times 2d)$, n is for n query, and each query has $(n'\times 2d)$ samples points.

and maybe the deformable attention should be written something like 
$A_{n \times n'} \cdot V_{I_{n \times n' \times 2d}}$  instead of $A_{n \times n'}\cdot V_{I_{n' \times 2d}}$ ?

**Comment by helq2612** [comment]
Oh yes, if you think the V is the one before points sampling, it is the same V for all the queries. For the deformable attention, each query get different Vs based on different sampled point sets (so that Vs are different for each query).

**Comment by Tsai-chia-hsiang** [comment]
"if you think the V is the one before points sampling, it is the same V for all the queries".

I think the $V_I$ from the eq(5) is AFTER the points sampling step. 

So I guess this code is using the original deformable detr that for each query gets a unique $V_q$ and the meaning of  eq(5) :  
$A_{n' \times n}\cdot V_{n'\times 2d}$ 

should be treated as 

$A_{n' \times n}\cdot V_{n\times n'\times 2d}$ 

makes better sense? 

Or this paper wants to use the same $V_I$ for all $q \in 1, 2, ... n$ that different from DEFORMABLE DETR: DEFORMABLE TRANSFORMERS
FOR END-TO-END OBJECT DETECTION ?

**Comment by helq2612** [comment]
Yes, you are right, we use the same V as original deformableDetr, so in eq5 it should be written as a summation of the dot product (over $n^\prime$ points for each query). Thank you for pointing out this!

**Comment by Tsai-chia-hsiang** [comment]
Ok, I see.
Thanks ~
